JDK updates and removals

emilazy opened the pull request:
## Description of changes

OpenJDK 23 is out, and 22 is end‐of‐life. This packages the new version for OpenJDK source builds, Temurin, and Zulu. I opted to drop the old versions entirely because these non‐LTS releases get released and go EOL every 6 months anyway so I don’t think it makes sense to carry around old derivations marked as insecure for versions with such short shelf lives. Sadly a few things aren’t ready for 23 yet (e.g. because of using the abandoned string template preview proposal), and have been marked as broken.

Some general thoughts on Nixpkgs JDK packaging from an outsider:

* OpenJDK source, Temurin, Semeru, Zulu, and Corretto feels like a needlessly high number of JDK distributions to me. Semeru has OpenJ9 and source builds have obvious value. I don’t see a good reason for us to have both Temurin and Zulu. It seems like we use Zulu on Darwin because it supports JavaFX and supports `aarch64-darwin` further back than Temurin does (because Temurin relies on upstream OpenJDK which [hasn’t backported `aarch64-darwin` support to JDK 8](https://github.com/adoptium/adoptium/issues/96)). I think if we could get the OpenJDK/OpenJFX source builds working on Darwin and bootstrap them from Temurin it would be a nicer situation overall; not supporting JDK 8 on `aarch64-darwin` seems okay. Zulu is pretty neglected, given its tedious manual update process; every Zulu JDK was out of date. If we stop using it on Darwin, I think it could reasonably be removed. (Switching from Temurin to Zulu for bootstrap might also be an option, if the derivation can be made more maintainable.) I’m not sure about the motivation for carrying Corretto; cc @rollf?

* 8, 11, 17, 21, and the latest version is… also a lot of versions to deal with. Backporting FFmpeg changes to old OpenJFXes was fairly painful. At least 8 will go EOL within the next couple years…

* There’s way too much copy and paste. The OpenJDK and OpenJFX derivations need deduplicating across versions, Temurin and Semeru are copy‐pasted from each other, the Linux and Darwin derivations of those could probably be deduplicated somewhat… if I end up looking at getting OpenJDK to build on Darwin I might try and do something about the duplication there.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux – `[ temurin-bin-8 temurin-bin-11 temurin-bin-17 temurin-bin-21 temurin-bin-23 temurin-jre-bin-8 temurin-jre-bin-11 temurin-jre-bin-17 temurin-jre-bin-21 temurin-jre-bin-23 openjdk8 moneydance jabref ] ++ lib.concatMap (jdk: [ jdk (jdk.override { enableJavaFX = true; }) ]) [ zulu8 zulu11 zulu17 zulu21 openjdk11 openjdk17 openjdk21 openjdk23 ]`
  - [x] aarch64-linux – `[ temurin-bin-8 temurin-bin-11 temurin-bin-17 temurin-bin-21 temurin-bin-23 temurin-jre-bin-8 temurin-jre-bin-11 temurin-jre-bin-17 temurin-jre-bin-21 temurin-jre-bin-23 openjdk8 openjdk11 moneydance jabref ] ++ lib.concatMap (jdk: [ jdk (jdk.override { enableJavaFX = true; }) ]) [ zulu8 zulu11 zulu17 zulu21 openjdk17 openjdk21 openjdk23 ]`
  - [ ] x86_64-darwin
  - [x] aarch64-darwin – `[ temurin-bin-11 temurin-bin-17 temurin-bin-21 temurin-bin-23 temurin-jre-bin-11 temurin-jre-bin-17 temurin-jre-bin-21 temurin-jre-bin-23 openjdk8 ] ++ lib.concatMap (jdk: [ jdk (jdk.override { enableJavaFX = true; }) ]) [ zulu8 zulu11 zulu17 zulu21 openjdk11 openjdk17 openjdk21 openjdk23 ]`
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy committed:
{openjdk,openjfx}23: init at 23-ga

emilazy committed:
zulu23: init at 23.0.0

emilazy committed:
temurin-{,jre-}bin-23: init at 23.0.0

emilazy committed:
moneydance: pin OpenJDK 23

emilazy committed:
{openjdk,openjfx}22: drop

emilazy committed:
jextract: pin JDK 23, mark as broken

emilazy committed:
jabref: pin JDK 21 and OpenJFX 23

This requires Java 21 but OpenJFX ≥ 22. As OpenJFX 22 is EOL,
we use OpenJFX 23, which technically only supports down to JDK 21,
but let’s hope for the best.

emilazy committed:
temurin-{,jre-}bin-22: drop

emilazy committed:
temurin-{,jre-}bin: temurin-{,jre-}bin-22 -> temurin{,jre-}bin-21

Pin the latest LTS version for consistency with OpenJDK.

emilazy committed:
temurin-{,jre-}bin-{8,11,17,21,22}: update

emilazy committed:
zulu22: drop

Actually, nobody even got around to making an alias for this one…

emilazy committed:
zulu17: 17.0.10 -> 17.0.12

emilazy committed:
zulu8: 8.0.{392,402} -> 8.0.422

emilazy committed:
zulu11: 11.0.22 -> 11.0.24

emilazy committed:
zulu21: 21.0.2 -> 21.0.4

emilazy committed:
cryptomator: pin JDK 23, mark as broken

emilazy commented:
Not going to fix the nixpkgs-vet warning given that it would result in inconsistency with all the other Zulu versions (and the Zulu packaging pattern wouldn’t work with `pkgs/by-name` anyway).

Infinidoge commented on pkgs/top-level/aliases.nix:
```diff
@@ -1163,8 +1163,13 @@ mapAliases {
   jdk20 = openjdk20; # Added 2024-08-01
   jdk20_headless = openjdk20; # Added 2024-08-01
+  openjdk22 = throw "OpenJDK 22 was removed as it has reached its end of life"; # Added 2024-09-24
+  openjdk22_headless = openjdk20; # Added 2024-09-24
```
```suggestion
  openjdk22_headless = openjdk22; # Added 2024-09-24
```

emilazy replied:
Whoops; fixed.

Infinidoge commented:
The OpenJDK 22 files still seem to be in the repository, might be an oversight.

rollf commented:
@emilazy Thank your for your feedback.

I have integrated Corretto because it is used in my organization. Several other people [had](https://github.com/NixOS/nixpkgs/issues/244456) [requested](https://github.com/NixOS/nixpkgs/issues/114795) it in the past. Several people have expressed the need for corretto on Darwin which currently doesn't exist. Here is a non-working/draft [PR](https://github.com/NixOS/nixpkgs/pull/290808) for some context.

In my case, the build infrastructure locks down the used Java version (via gradle, `vendor = JvmVendorSpec.AMAZON`) so I could not build anything (on NixOS) without removing that condition (and using OpenJDK).

One thing I'd like to explore in the future is separate the corretto build from the OpenJDK builds. Currently, `mk-corretio.nix` overridesAttributes on the corresponding JDK build. I think (now) that this is needlessly binding the two projects.

thiagokokada commented:
> OpenJDK source, Temurin, Semeru, Zulu, and Corretto feels like a needlessly high number of JDK distributions to me.

I concur, unless there is a maintainer committed to maintaining those variations I would also be in favor of just removing them as much as possible. Otherwise they will just bit rot (e.g. stop building) or be a security issue since they get terrible out of date.

In my opinion we need to maintain the minimal amount of JDKs possible: so if we need e.g. Zulu JDK8 to bootstrap other JDK for Darwin let's just keep that specific version (this is an example, not sure how Darwin bootstrap works). And ideally keep only OpenJDK as the way to go for Java here in nixpkgs.

rollf commented:
> I concur, unless there is a maintainer committed to maintaining those variations I would also be in favor of just removing them as much as possible. Otherwise they will just bit rot (e.g. stop building) or be a security issue since they get terrible out of date.

I am committed to maintain Corretto on Linux. I use this daily. I have updated it regularly ([here](https://github.com/NixOS/nixpkgs/pull/342262) is my latest approved-but-not-merged-PR).

Tungsten842 commented:
I believe that jdk21, jdk17, jdk11 and jdk8 require an update as well. For each of these a new version is available, and they have not been updated in quite a while.

emilazy commented:
> The OpenJDK 22 files still seem to be in the repository, might be an oversight.

Yeah, rebase calamity. Fixed.

> I have integrated Corretto because it is used in my organization. Several other people [had](https://github.com/NixOS/nixpkgs/issues/244456) [requested](https://github.com/NixOS/nixpkgs/issues/114795) it in the past. Several people have expressed the need for corretto on Darwin which currently doesn't exist. Here is a non-working/draft [PR](https://github.com/NixOS/nixpkgs/pull/290808) for some context.
> 
> In my case, the build infrastructure locks down the used Java version (via gradle, `vendor = JvmVendorSpec.AMAZON`) so I could not build anything (on NixOS) without removing that condition (and using OpenJDK).

Fair enough. It’s a bit weird to me that people are pinning requirements for what are, to my understanding, basically builds of the same source code with a handful of patches on top, but clearly there’s some benefit to having it. Since they only ship LTS releases and you’re keeping it cared for, it didn’t cause me any maintenance burden here, and so I have no particular motivation to push for its removal :) Thanks for putting in the maintenance work.

> In my opinion we need to maintain the minimal amount of JDKs possible: so if we need e.g. Zulu JDK8 to bootstrap other JDK for Darwin let's just keep that specific version (this is an example, not sure how Darwin bootstrap works). And ideally keep only OpenJDK as the way to go for Java here in nixpkgs.

Using Zulu to bootstrap our source OpenJDK 8 on `aarch64-darwin` wouldn’t help that much, as the OpenJDK 8 source itself needs extensive patching to work with that platform. There are public patches for this, to my understanding, but it seems like upstreaming them has stalled out. I don’t think we should try and maintain a port of a version that is going to be EOL in a year or two to `aarch64-darwin` ourselves. We should just pick one of Temurin or Zulu, use it for bootstrap of source OpenJDKs, ship the source build on all platforms, and let `jdk8` be broken on `aarch64-darwin`, as it is upstream.

In fact, I would personally lean towards just dropping JDK 8 sooner. Debian already has, for instance.

> I believe that jdk21, jdk17, jdk11 and jdk8 require an update as well. For each of these a new version is available, and they have not been updated in quite a while.

And this is why shipping five versions of the JDK with the maintainer resources we seem to have is a pain…

I’ll see if I can bump them.

emilazy committed:
openjdk: remove unused bootstrap files

These will still be available in the Git history if anyone wants to
try resurrecting a source‐based OpenJDK bootstrap in the future. In
the meantime they’re just confusing to have around unreferenced.

emilazy committed:
openjdk11: 11.0.23+9 -> 11.0.24+8

emilazy committed:
openjdk21: remove from `info.json`

The derivation doesn’t use this, so it’s confusing and misleading
to have it in there.

emilazy committed:
openjdk17: 17.0.11+9 -> 17.0.12+7

emilazy committed:
openjdk8: 8u412-ga -> 8u422-ga

emilazy committed:
openjdk21: 21.0.3+9 -> 21.0.4+7

thiagokokada commented:
> Using Zulu to bootstrap our source OpenJDK 8 on aarch64-darwin wouldn’t help that much, as the OpenJDK 8 source itself needs extensive patching to work with that platform. There are public patches for this, to my understanding, but it seems like upstreaming them has stalled out. I don’t think we should try and maintain a port of a version that is going to be EOL in a year or two to aarch64-darwin ourselves. We should just pick one of Temurin or Zulu, use it for bootstrap of source OpenJDKs, ship the source build on all platforms, and let jdk8 be broken on aarch64-darwin, as it is upstream.

Oh no, I didn't mean to build OpenJDK 8 for aarch64-darwin from Zulu 8, that was an example (e.g.: if we need any JDK8 to bootstrap the remaining OpenJDK versions that we care, we could use Zulu 8 and keep only that in nixpkgs, dropping the other versions). I completely agree that we should not care about having a working OpenJDK 8 build for aarch64-darwin, unless there is official support in upstream.

> In fact, I would personally lean towards just dropping JDK 8 sooner. Debian already has, for instance.

That I am not completely sure. If there is no packages in nixpkgs that depends on JDK 8 I think it is fine, but there are probably some users of Java 8 out there. I don't think there are a lot of them though, but it is difficult to have any idea of package usage in nixpkgs.

thiagokokada reviewed (approved):
LGTM.

emilazy commented:
```
configure: error: "ga" is not a valid numerical value for VERSION_BUILD
```

Sigh.

thiagokokada commented on pkgs/development/compilers/openjdk/17.nix:
```diff
@@ -93,7 +93,7 @@ let
     configureFlags = [
       "--with-boot-jdk=${openjdk-bootstrap.home}"
-      "--with-version-build=${version.build}"
+      #"--with-version-build=${version.build}"
```
Can you just confirm that the reported version is correct without setting this?

emilazy replied:
I’m still building :)

But FWIW the other `-ga` versions, including 22 before this PR, don’t pass this flag, and the Temurin build system omits it for `-ga`, so I believe this is correct.

emilazy replied:
We could just use the `+7` version rather than `-ga` since it’s the same tag, I guess. I don’t really know what’s best here.

thiagokokada replied:
> We could just use the +7 version rather than -ga since it’s the same tag, I guess. I don’t really know what’s best here.

I think I am more comfortable to switch to the non-ga versions, but I am fine for whatever you decide here.

emilazy replied:
OK, done. Didn’t touch 8 or 23 since they were already like that. We should just be consistent with Temurin but it’ll have to wait until after refactoring. It’s too many builds for me to bother running again, but this should be consistent with the previous state.

emilazy commented:
All the builds went through. This should be ready, modulo version stuff.

thiagokokada reviewed (approved):
LGTM.

thiagokokada commented:
Eval error, can you take a look @emilazy?

thiagokokada commented:
@ofborg eval

thiagokokada commented:
> Eval error, can you take a look @emilazy?

Seems to be a transient issue. `@ofborg eval` worked.

jlesquembre commented on pkgs/by-name/je/jextract/package.nix:
```diff
@@ -46,8 +46,10 @@ stdenv.mkDerivation {
+    platforms = jdk23.meta.platforms;
     license = licenses.gpl2Only;
     maintainers = with maintainers; [ jlesquembre sharzy ];
+    # Not yet updated for JDK 23
+    broken = true;
```
It would be more respectful to jextract users to wait with the jdk22 removal

thiagokokada replied:
I think keeping older versions of Java that are possible insecure is more of an issue than having one specific package to break. We could just have marked the JDK22 (and all packages that depend on it) as insecure instead and keep it for longer, but I concur that we should, if anything, to reduce the number of Java packages in the tree to make sure that the ones that we have are up to date.

emilazy replied:
The `jextract` for the LTS JDK 21 is still in‐tree. There’s been seemingly no movement in the `jextract` repository porting it to an on‐EOL JDK, so I don’t know what’s going on there. The same upstream that maintains `jextract` sets the OpenJDK support policy and since it is guaranteed that we will get these versions going EOL every release (when it’s not time for a new LTS) we can’t let the list of insecure, unsupported JDKs keep growing indefinitely, as happened before the 24.11 cycle. I would prefer we move to a scheme like Fedora where the latest JDK is available as `openjdk_latest` and simply rolls over to the new version when released.

msgilligan replied:
> older versions of Java that are possibly insecure

Well all software is possibly insecure. I know that you mean unmaintained, which is similar.

> We could just have marked the JDK22 (and all packages that depend on it) as insecure instead and keep it for longer

That would seem like a better solution to me.

> reduce the number of Java packages in the tree to make sure that the ones that we have are up to date

I agree with this as well. But I would suggest something like what I see for [java in Debian Trixie](https://packages.debian.org/trixie/java/) where they support:

* 17
* 21
* 22
* 23
* 24-ea

At an absolute minimum provide: LTS versions, the latest release, the previous release. It would also be nice to see EA versions installable through Nix, but obviously someone has to do the work.

Releases that are unmaintained upstream could be marked as unsupported/insecure/deprecated, just so long as projects aren't broken without a few months warning.

In the case of `jextract`, which is a tool to build bindings for Java's Foreign Function & Memory API which went final in JDK 22. The only way (I'm pretty sure) to build bindings for the final API is with the JDK 22-based jextract. So, at this point, there is no way to do that using nixpkgs.

msgilligan replied:
> The `jextract` for the LTS JDK 21 is still in‐tree.

As I mentioned in my reply above, this version is not very useful because it is for a non-final FFM API and requires use of JDK 21 with "preview features" enabled.

> There’s been seemingly no movement in the `jextract` repository porting it to a non‐EOL JDK

This appears to be the case and is disappointing. Maybe someone should reach out to the OpenJDK project and ask about it.

> we can’t let the list of insecure, unsupported JDKs keep growing indefinitely

I agree. But hopefully there's a way to do this with a little more warning (e.g. a 3-6 months) to dependent projects.

emilazy replied:
Honestly if we wanted to diverge from the Fedora policy of LTSes + rolling latest release then I’d prefer simply skipping the 6‐month releases entirely. They go EOL like clockwork so software choosing to live on the bleeding edge should be prepared to roll over to the next release when it comes out – that’s the price of relying on features in releases with a very limited support lifespan. Why a first‐party project like `jextract` can’t keep up with that, I don’t know; you’d presumably have to ask the OpenJDK developers. It’s already a huge pain for Nixpkgs to deal with software with such a limited lifespan that isn’t perfectly synchronized to our release cycle.

Note that we usually carry only one version of software in Nixpkgs; OpenJDK is already very high up the list of packages we carry a large number of versions for, with e.g. more versions than Node.js and the same number as CPython 3. We mark EOL language implementations as insecure and consider them high priorities for removal because they are highly complex pieces of software with a huge attack surface that has a downstream security impact on every package that uses them. Just look at the [OpenJDK Vulnerability Advisories](https://openjdk.org/groups/vulnerability/advisories/) to see how regularly security issues are found and fixed. EOL JDKs don’t get those fixes.

thiagokokada replied:
> Well all software is possibly insecure. I know that you mean unmaintained, which is similar.

Well, looking at the latest advisory for Java it seems there are new security issues affecting all supported Java versions. It is very unlikely this doesn't also affect JDK 22 so we actually know it has security issues: https://openjdk.org/groups/vulnerability/advisories/2024-10-15.

Those security advisories are released every 3 months and almost always have at least one security issue, so for a version that is maintained by 9 months like our stable branches we are unlikely to ship the previous version and keep it secure.

I actually like the proposal from @emilazy here to keep all LTS versions+`openjdk_latest` with the last version, because at least we can ensure that it is possible to keep our JDK versions up to date in stable.

emilazy replied:
FWIW I spent about 30 minutes looking into a way to patch `jextract` for the latest JDK and I didn’t succeed, but I suspect someone with more familiarity would have an easier time with it. It didn’t seem like it should be impossible by any means; it just exceeded the time‐box I had allocated to it. If we could do that then it could also be sent upstream to OpenJDK.

> I agree. But hopefully there's a way to do this with a little more warning (e.g. a 3-6 months) to dependent projects.

The OpenJDK release and support cycle is fixed – the 6 months warning is when a new non‐LTS release comes out, and the 3 months warning is 3 months after that. That cycle is on the OpenJDK project; we don’t have the resources to support them longer than upstream does. We’ve regularly lagged behind on security‐critical updates even to the versions we currently have, and 24.05 was in a much worse state (I think all the JDKs there should actually be marked as `knownVulnerabilities`).

I believe our stable branches are actually maintained for 7 months? But: that means that OpenJDK 23 is going to go EOL during the 24.11 cycle, which is really awkward. Either we mark it with `knownVulnerabilities` and have only programs using LTS JDKs working without a security version for the last few months of the release cycle, or we backport a bump to the next JDK and potentially break other software in the process. I believe long‐term support distributions like RHEL don’t even bother shipping non‐LTS JDKs. It’s really something you should only depend on if you can be sufficiently agile at supporting upcoming versions, given the upstream policy.

emilazy replied:
[This mailing list message](https://mail.openjdk.org/pipermail/jextract-dev/2024-September/001948.html) claims that `jextract` should basically “just work” with JDK 23. That was unfortunately not my experience, but it could just be build system issues I’d be running into, so if someone wanted to have a go at getting the package unbroken I suspect it shouldn’t be too difficult.

msgilligan replied:
The arguments about security are compelling. And I know this is a volunteer project with limited resources. I am new to Nixpkgs and am trying to contribute as much as I can (e.g. https://github.com/NixOS/nixpkgs/pull/271127), but am not (yet) ready to try to (help) support OpenJDK itself.

I have been trying to replace (as much as possible) Homebrew and [SDKMAN!](https://sdkman.io) with Nixpkgs.

I personally prefer to focus on getting `jextract` updated: https://github.com/NixOS/nixpkgs/pull/271127

But I also want to share my experiences a a Nixpkgs user and having something like `jdk22` disappear so quickly is not very friendly to downstream projects. A difference between installing JDK binaries with Homebrew or SDKMAN! and using Nix flakes is that dropping JDK 22 will actually break Nix builds, whereas Homebrew or SDKMAN! will allow you to leave your old (unmaintained/insecure) JDK installed and usable.

emilazy replied:
FWIW, it’s possible to pin the last Nixpkgs version with OpenJDK 22, and use that package alongside other packages on your system, thanks to Nix’s hermetic pinning of dependency trees. That’s of course inconvenient, and in some ways slightly worse than if we were carrying an insecure‐marked OpenJDK 22 in the current tree (namely: its dependencies won’t get updates either, though since the JDK itself is probably the biggest source of vulnerabilities in its dependency closure that might not matter too much), but hopefully it can at least serve at a pinch when you really can’t do without the old version. (It also avoids having to build the JDK yourself, since we don’t spend Hydra build farm resources on insecure packages.)

I realize it’s frustrating from a user point of view, though, and I sympathize with the pain of dealing with short support cycles – ideally OpenJDK would provide an overlapping support window for their non‐LTS releases, but they don’t. I think they would have the ability to do that; we support a one month security support overlap between our own stable versions despite having considerably fewer resources than Oracle. I don’t really know why they don’t; I was wondering if perhaps it’s because they sell extended commercial support, but that’s only for the LTS releases. I get the impression that they consider their non‐LTS releases to essentially be previews for the next LTS.

If you want to try updating `jextract` to the latest early access build and seeing if it’ll work on JDK 23 that would be great, and please do ping me for review! I can try to help if you run into any issues as well. I didn’t want to mark it as broken here but making sure we had the latest supported JDK and weren’t at risk of shipping an EOL one in 24.11 took priority. If we could get it fixed before the 24.11 release I’d be very happy.

msgilligan commented:
From the perspective of getting newer JDKs (e.g. 25) including in Nixpkgs sooner, I prioritized

* openjdk -- the upstream Open Source,  reference build for JDK
* temurin-bin -- popular Open Source pre-built distribution which Nixpkgs uses to bootstrap OpenJDK
* zulu -- popular Open Source pre-built distribution with JavaFX. On Darwin, Nixpkgs aliases `jdk` to `zulu`
* graalvm-ce -- Open Source "community edition" of GraalVM that provides the `native-image` compiler
* graalvm-oracle -- non-free GraalVM, but provides binaries during Early Access


From https://github.com/NixOS/nixpkgs/issues/430389#issuecomment-3304698501
